File: Selenium/Project/Activity9.java
/*Traversing tables 2
        Goal: Open the leads page and print the usernames and full names from the table*/
import io.github.bonigarcia.wdm.WebDriverManager;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.firefox.FirefoxDriver;
import org.openqa.selenium.interactions.Actions;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;
import org.testng.annotations.AfterClass;
import org.testng.annotations.BeforeClass;
import org.testng.annotations.Test;

import java.time.Duration;
import java.util.List;

public class Activity9 {
    // Declare the WebDriver object
    private WebDriver driver;
    WebDriverWait wait;
    @BeforeClass
    public void setupMethod() {
        // Set up the Firefox driver
        WebDriverManager.firefoxdriver().setup();
        //Create a new instance of the Firefox driver
        driver = new FirefoxDriver();
        //Open browser and navigate to url
        driver.get("https://alchemy.hguy.co/crm");
    }

    @Test
    public void traversingTables2() throws InterruptedException {
        //Login the application

        // Find the username field and enter the username
        driver.findElement(By.xpath("//input[@id='user_name']")).sendKeys("admin");
        // Find the password field and enter the password
        driver.findElement(By.xpath("//input[@id='username_password']")).sendKeys("pa$$w0rd");
        // Find the login button and click it
        driver.findElement(By.xpath("//input[@name='Login']")).click();
        String expectedTitle = "SuiteCRM";
        System.out.println("Title of the WebPage is:" + driver.getTitle());
        if (driver.getTitle() != null && driver.getTitle().contains(expectedTitle)) {
            System.out.println("Home page is opened");
        } else {
            System.out.println("Home page could not open.");
        }

        //Find the salesMenuItem
        wait = new WebDriverWait(driver, Duration.ofSeconds(10));
        WebElement salesMenuItem = driver.findElement(By.xpath("//a[@id='grouptab_0']"));
        // object of Actions with method moveToElement
        Actions a = new Actions(driver);
        a.moveToElement(salesMenuItem).perform();

        //identify sub-menu element Accounts
        WebElement leadsSubMenu = driver.
                findElement(By.xpath("//*[text()='Leads']"));
        //move to element and click
        a.moveToElement(leadsSubMenu).click().perform();
        By leadsHeaderPresent = By.xpath("//h2[@class='module-title-text']");
        wait.until(ExpectedConditions.visibilityOfElementLocated(leadsHeaderPresent));


        // Finding number of rows in a web table.
        List allRows = driver.findElements(By.xpath("//table[@class='list view table-responsive']/tbody/tr"));
        System.out.println("Total data rows found in table:" + (allRows.size()));

        // print the first 10 values in the Name column and the User column of the table to the console
        System.out.println("print the first 10 values in the Name column and the User column of the table to the console:");
        System.out.println("Name Column" +" "+"|"+" "+"User Column");
        for (int i = 1; i <= 10; i++) {
            //Find the name column values and print it
            WebElement nameColumn = driver.findElement(By.xpath("//table[@class='list view table-responsive']/tbody/tr[" + i + "]/td[3]"));
            //Find the user column values and print it
            WebElement userColumn = driver.findElement(By.xpath("//table[@class='list view table-responsive']/tbody/tr[" + i + "]/td[8]"));
            System.out.print(nameColumn.getText() + " |" +userColumn.getText());
            System.out.println();
            }
        }


    @AfterClass
    public void afterMethod() {
        //Close the browser
        driver.close();
    }
}

File: Selenium/Project/Activity9.py
#Traversing tables 2
#Goal: Open the leads page and print the usernames and full names from the table.


# Import webdriver from selenium
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.firefox.service import Service as FirefoxService
from webdriver_manager.firefox import GeckoDriverManager
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.common.exceptions import TimeoutException
from selenium.webdriver import ActionChains

# Set up the Firefox Driver with WebDriverManger
service = FirefoxService(GeckoDriverManager().install())

# Start the Driver
with webdriver.Firefox(service = service) as driver:
  
    # Open the browser to the URL
    driver.get("https://alchemy.hguy.co/crm")

    # Read the title of the website and verify the text
    print("Home page title:"+driver.title)
  
    # Find the username field and enter the username
    driver.find_element(By.XPATH,"//input[@id='user_name']").send_keys("admin")
    # Find the password field and enter the password
    driver.find_element(By.XPATH,"//input[@id='username_password']").send_keys("pa$$w0rd")
    # Find the login button and click it
    driver.find_element(By.XPATH,"//input[@name='Login']").click()
  
    # Wait for a maximum of 10 seconds
    wait = WebDriverWait(driver, 10)  

    # Find the salesMenuItem
    sales_menu_item = driver.find_element(By.XPATH, "//a[@id='grouptab_0']")

    # Create an instance of Actions
    actions = ActionChains(driver)
    actions.move_to_element(sales_menu_item).perform()

    # Identify sub-menu element Leads
    leads_sub_menu = driver.find_element(By.XPATH, "//*[text()='Leads']")

    # Move to element and click
    actions.move_to_element(leads_sub_menu).click().perform()

    leads_header_present = By.XPATH, "//h2[@class='module-title-text']"
    wait.until(EC.visibility_of_element_located(leads_header_present))

    # Finding number of rows in a web table.
    all_rows = driver.find_elements(By.XPATH, "//table[@class='list view table-responsive']/tbody/tr")
    print("Total data rows found in table: " + str(len(all_rows)))

    # Print the first 10 values in the Name column and the User column of the table to the console.
    print("Print the first 10 values in the Name column and the User column of the table to the console:")
    print("Name Column | User Column")
    for i in range(1, 11):
        # Find the name column values and print them
        name_column = driver.find_element(By.XPATH, "//table[@class='list view table-responsive']/tbody/tr[" + str(i) + "]/td[3]")
        # Find the user column values and print them
        user_column = driver.find_element(By.XPATH, "//table[@class='list view table-responsive']/tbody/tr[" + str(i) + "]/td[8]")
        print(name_column.text + " | " + user_column.text)    


    # Close the browser
    driver.close()